programs/clearing_house/src/state/market.rs

```diff
@@ -44,12 +52,64 @@ pub struct Market {
 }
 
 impl Market {
-    pub fn get_margin_ratio(&self, margin_type: MarginRequirementType) -> u32 {
-        match margin_type {
-            MarginRequirementType::Initial => self.margin_ratio_initial,
-            MarginRequirementType::Partial => self.margin_ratio_partial,
-            MarginRequirementType::Maintenance => self.margin_ratio_maintenance,
-        }
+    pub fn get_margin_ratio(
+        &self,
+        size: u128,
+        margin_type: MarginRequirementType,
+    ) -> ClearingHouseResult<u32> {
+        let margin_ratio = match margin_type {
+            MarginRequirementType::Initial => max(
+                self.margin_ratio_initial as u128,
+                calculate_size_markup_liability_weight(
+                    size,
+                    self.imf_factor,
+                    self.margin_ratio_partial as u128,
+                    MARGIN_PRECISION,
+                )? + 1,
+            ),
+            MarginRequirementType::Partial => calculate_size_markup_liability_weight(
+                size,
+                self.imf_factor,
+                self.margin_ratio_partial as u128,
+                MARGIN_PRECISION,
+            )?,
+            MarginRequirementType::Maintenance => self.margin_ratio_maintenance as u128,
+        };
+
+        Ok(margin_ratio as u32)
+    }
+
+    pub fn get_unsettled_asset_weight(
+        &self,
+        unsettled_pnl: i128,
+        margin_type: MarginRequirementType,
+    ) -> ClearingHouseResult<u128> {
+        // the asset weight for a position's unrealized pnl + unsettled pnl in the margin system
+        // > 0 (positive balance)
+        // < 0 (negative balance) always has asset weight = 1
+        let unsettled_asset_weight = if unsettled_pnl > 0 {
+            // todo: only discount the initial margin s.t. no one gets liquidated over upnl?
```

Comment:
what would that look like?

Reply:
user has two positions one long one short

the long position is up big and the short position is down big. if you discount that upnl heavily liquidations then they could get liq'd. (i guess as long as it was for large sizes only its not so bad to do for partial liquidations)

Reply:
so does that only mean used the asset weight for initial margin?